File: framework/languages/cpp/http-client/src/runtime/cookie_jar.cpp
/* SPDX-License-Identifier: Apache-2.0 */

#include "runtime/cookie_jar.hpp"

#include "runtime/text.hpp"

#include <algorithm>
#include <optional>

namespace zlink::http_client::detail
{

namespace
{

constexpr std::size_t max_cookies_per_host = 128;

bool path_matches_cookie (const std::string &request_path, const std::string &cookie_path)
{
    if (cookie_path.empty () || cookie_path == "/") {
        return true;
    }
    if (request_path == cookie_path) {
        return true;
    }
    if (request_path.rfind (cookie_path, 0) != 0) {
        return false;
    }
    return cookie_path.back () == '/' || request_path[cookie_path.size ()] == '/';
}

} // namespace

void cookie_jar_t::store (const std::string &host, const std::string &set_cookie_header)
{
    std::string_view remaining (set_cookie_header);
    const auto take_segment = [&remaining] () -> std::optional<std::string> {
        if (remaining.empty ()) {
            return std::nullopt;
        }
        const auto semicolon = remaining.find (';');
        auto segment = trim (remaining.substr (0, semicolon));
        remaining = semicolon == std::string_view::npos ? std::string_view ()
                                                        : remaining.substr (semicolon + 1);
        return segment;
    };

    const auto pair = take_segment ();
    if (!pair) {
        return;
    }
    const auto equals = pair->find ('=');
    if (equals == std::string::npos) {
        return;
    }
    cookie_t cookie{.host = host,
                    .name = trim (std::string_view (*pair).substr (0, equals)),
                    .value = trim (std::string_view (*pair).substr (equals + 1)),
                    .path = "/",
                    .secure = false};
    if (cookie.name.empty ()) {
        return;
    }

    bool expired = false;
    while (auto attribute = take_segment ()) {
        const auto attr_equals = attribute->find ('=');
        const auto name = attr_equals == std::string::npos
                            ? *attribute
                            : trim (std::string_view (*attribute).substr (0, attr_equals));
        const auto value = attr_equals == std::string::npos
                             ? std::string ()
                             : trim (std::string_view (*attribute).substr (attr_equals + 1));
        if (iequals (name, "path") && !value.empty ()) {
            cookie.path = value;
        } else if (iequals (name, "secure")) {
            cookie.secure = true;
        } else if (iequals (name, "max-age")) {
            try {
                expired = std::stol (value) <= 0;
            }
            catch (const std::exception &) {
            }
        }
    }

    const std::lock_guard<std::mutex> lock (_mutex);
    _cookies.erase (std::remove_if (_cookies.begin (), _cookies.end (),
                                    [&] (const cookie_t &existing) {
                                        return existing.host == host && existing.name == cookie.name
                                               && existing.path == cookie.path;
                                    }),
                    _cookies.end ());
    if (!expired) {
        _cookies.push_back (std::move (cookie));
        std::size_t cookies_for_host = 0;
        for (const auto &stored : _cookies) {
            if (stored.host == host) {
                ++cookies_for_host;
            }
        }
        while (cookies_for_host > max_cookies_per_host) {
            auto oldest = std::find_if (_cookies.begin (), _cookies.end (),
                                        [&] (const cookie_t &stored) {
                                            return stored.host == host;
                                        });
            if (oldest == _cookies.end ()) {
                break;
            }
            _cookies.erase (oldest);
            --cookies_for_host;
        }
    }
}

std::string
cookie_jar_t::header_for (const std::string &host, const std::string &path, bool secure) const
{
    const std::lock_guard<std::mutex> lock (_mutex);
    std::string header;
    for (const auto &cookie : _cookies) {
        if (cookie.host != host || (cookie.secure && !secure)
            || !path_matches_cookie (path, cookie.path)) {
            continue;
        }
        if (!header.empty ()) {
            header += "; ";
        }
        header += cookie.name + "=" + cookie.value;
    }
    return header;
}

} // namespace zlink::http_client::detail

File: framework/languages/node/packages/http-client/src/runtime/cookie-jar.ts
/* SPDX-License-Identifier: Apache-2.0 */

/**
 * Wrapper-owned cookie jar mirroring the deliberately narrow C++ `cookie_jar.cpp` semantics so
 * behaviour matches across languages: host-exact storage (no `Domain`), default `Path=/`, only
 * `Path`/`Secure`/`Max-Age<=0` honoured (`Domain`/`Expires` ignored), at most 128 cookies per host,
 * and secure cookies sent only on secure requests. `fetch`/undici provide no server-side persistent
 * cookie jar, so this is implemented in the wrapper.
 */

interface Cookie {
  host: string;
  name: string;
  value: string;
  path: string;
  secure: boolean;
}

const MAX_COOKIES_PER_HOST = 128;

export class CookieJar {
  private readonly cookies: Cookie[] = [];

  store(host: string, setCookieHeader: string): void {
    const segments = setCookieHeader.split(';');
    const pair = (segments.shift() ?? '').trim();
    const equals = pair.indexOf('=');
    if (equals < 0) {
      return;
    }

    const name = pair.slice(0, equals).trim();
    const value = pair.slice(equals + 1).trim();
    if (name.length === 0) {
      return;
    }

    let path = '/';
    let secure = false;
    let expired = false;
    for (const raw of segments) {
      const attribute = raw.trim();
      const attrEquals = attribute.indexOf('=');
      const attrName = (attrEquals < 0 ? attribute : attribute.slice(0, attrEquals)).trim();
      const attrValue = attrEquals < 0 ? '' : attribute.slice(attrEquals + 1).trim();
      if (attrName.toLowerCase() === 'path' && attrValue.length > 0) {
        path = attrValue;
      } else if (attrName.toLowerCase() === 'secure') {
        secure = true;
      } else if (attrName.toLowerCase() === 'max-age') {
        const maxAge = Number.parseInt(attrValue, 10);
        if (!Number.isNaN(maxAge)) {
          expired = maxAge <= 0;
        }
      }
    }

    for (let i = this.cookies.length - 1; i >= 0; i--) {
      const existing = this.cookies[i];
      if (existing.host === host && existing.name === name && existing.path === path) {
        this.cookies.splice(i, 1);
      }
    }

    if (expired) {
      return;
    }

    this.cookies.push({ host, name, value, path, secure });
    let countForHost = this.cookies.filter((stored) => stored.host === host).length;
    while (countForHost > MAX_COOKIES_PER_HOST) {
      const oldest = this.cookies.findIndex((stored) => stored.host === host);
      if (oldest < 0) {
        break;
      }
      this.cookies.splice(oldest, 1);
      countForHost--;
    }
  }

  headerFor(host: string, path: string, secure: boolean): string {
    const parts: string[] = [];
    for (const cookie of this.cookies) {
      if (cookie.host !== host || (cookie.secure && !secure) || !pathMatches(path, cookie.path)) {
        continue;
      }
      parts.push(`${cookie.name}=${cookie.value}`);
    }
    return parts.join('; ');
  }
}

function pathMatches(requestPath: string, cookiePath: string): boolean {
  if (cookiePath.length === 0 || cookiePath === '/') {
    return true;
  }
  if (requestPath === cookiePath) {
    return true;
  }
  if (!requestPath.startsWith(cookiePath)) {
    return false;
  }
  return cookiePath.endsWith('/') || requestPath[cookiePath.length] === '/';
}
